### ISSUE TITLE
[feature request] Don't make requests to domains beyond the target URL

### ISSUE THREAD

**Issue by ar-rm** [zolrath/obsidian-auto-link-title#68]
Hi,

Obsidian Auto Link Title is one of my favorite and most used Obsidian plugins. Thank you very much!!

### tl;dr:

- OALT makes outgoing connections to domains **linked** by the page at the requested URL.
- It would be best if it didn't, since those domains shouldn't be necessary to obtain the `title`.
- Proposed solution: Force [BrowserWindow](https://www.electronjs.org/docs/latest/api/browser-window) to download the requested URL's `html` only and go no further.

### Overview:

The only downside I experience regularly is the high number of **unnecessary** outgoing connections as a result of the resources linked in the parent url's `html`.

(Or at least that's what it *seems* like is happening. But I could be wrong.)

[Little Snitch](https://www.obdev.at/products/littlesnitch/) (software firewall) warns me about all of the extra outgoing requests when I run OALT.

Domains like:

cdn.`[xx]`.`[tld]`
`[xx]`.cloudflare.com
`[xx]`.jsdelivr.com
s`[n]`.amazon.com
github.githubassets.com
cdn.affiliatable.io
geniuslinkcdn.com
videoplayerhub.com
chimpstatic.com

In addition, there are all of the domains associated with tracking beacons/pixels, analytics, etc..

Unless I'm missing something, there's **no** need to make requests to those far off sites just to retrieve the contents of the `<title>` tag. That's just wasted data/bandwidth*/energy at best, and has privacy/security implications at worst.

(*Hopefully those external links are only triggering a simple html request [via `BrowserWindow`] and the non-html resources aren't being download to nowhere.)

--

The first alert I receive from Little Snitch is related to the actual site, which is fine, and I allow it. But the subsequent ones are usually totally unnecessary.

I can usually **deny** them all (one-at-a-time) and the `title` is ultimately retrieved. Sometimes not though*. In those cases, when OALT returns `Site Unreachable`, I usually just manually fetch the title (like a caveman :-).

(*Unfortunately I don't immediately know when I can safely start rejecting *everything*. Sometimes OALT seems to have retrieved the `title`, but the text in Obsidian remains as `Fetching Title#abcd`, while I continue to accept/deny the various alerts from Little Snitch. I assume this has something to do with JS and main-thread blocking.)

This game of Whac-A-Mole will slow over time (as I create rules that apply to multiple sites, ie: `https://www.googletagmanager.com`), but there's an endless number of these servers.

### Propose:

- Don't follow links to ANY *external* resources beyond the requested URL.
- The contents of `<title>` should be in the initial `fetch` and then OALT doesn't need to follow any other links. And if not, following something like a `301` is not the same as following all of the resources linked in the `<head>` (assuming that's what's happening).

### Possible implementations:

I see what you wrote about using Electron's [BrowserWindow](https://www.electronjs.org/docs/latest/api/browser-window) module.

I looked trough its [api documentation](https://github.com/electron/electron/blob/main/docs/api/browser-window.md#new-browserwindowoptions) and didn't see anything that jumped out at me re: `filtering`, `no-follow`, etc.. But I'm not an Electron expert.

I searched a bit more and found the following, which recommends using an additional package called [NightmareJS](https://github.com/segmentio/nightmare), which looks like it can provide the necessary filtering:

https://stackoverflow.com/questions/40705745/nightmarejs-how-to-block-loading-resources

Another author even chimed in, with their package `nightmare-load-filter`, which "Adds pre-load content filtering to your Nightmare scripts":

https://github.com/rosshinkley/nightmare-load-filter

Of course, there might be a better way to implement this. But I figured I should report what I found in case it could give you a head start in implementing this.

Thanks!
